Added the CD3DX12FeatureSupport class

Sanithovski opened the pull request:
- Added a class to handle feature support checks
- Implemented half of the feature checks
- Added D3D12_HIGHEST_SHADER_MODEL and D3D12_HIGHEST_ROOT_SIGNATURE_VERSION to help identify the highest versions in the d3d12.h header
- Added a test (feature_check_test.cpp) to ensure that the code compiles and is functioning properly (not a rigorous test case)
- Modified CMakeLists to include the new test

Sanithovski commented on include/directx/d3d12.h:
```diff
@@ -2257,7 +2257,9 @@ enum D3D_ROOT_SIGNATURE_VERSION
-        D3D_ROOT_SIGNATURE_VERSION_1_1	= 0x2
+        D3D_ROOT_SIGNATURE_VERSION_1_1	= 0x2,
+        
+        D3D_HIGHEST_ROOT_SIGNATURE_VERSION = D3D_ROOT_SIGNATURE_VERSION_1_1 // TODO: Update main repo
```
Added a "HIGHEST_VERSION" entry to help identify the highest version

shawnhar replied:
Adding this as a parallel with D3D_HIGHEST_SHADER_MODEL makes sense to me.  My only slight doubt is that shader models change often (usually with every new D3D update) while root signature has only ever changed once, and might not again.  What does @amarpMSFT think, is it reasonable to add a max value to this enum?

Sanithovski commented on include/directx/d3d12.h:
```diff
@@ -2300,7 +2302,8 @@ enum D3D_SHADER_MODEL
         D3D_SHADER_MODEL_6_6	= 0x66,
-        D3D_SHADER_MODEL_6_7	= 0x67
+        D3D_SHADER_MODEL_6_7	= 0x67,
+        D3D_HIGHEST_SHADER_MODEL = D3D_SHADER_MODEL_6_7 // TODO: Pull from latest d3d12.h
```
Added a "HIGHEST_VERSION" entry to help identify the highest version. Note: I find this entry to be present in the internal D3D12 library header but not in the open source repo. Probably updated recently?

shawnhar replied:
Good find!  This is indeed a recent addition that will make it across into the public headers the next time we update it.

Sanithovski commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,277 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+    FEATURE_SUPPORT_GET(BOOL, m_dOptions, VPAndRTArrayIndexFromAnyShaderFeedingRasterizerSupportedWithoutGSEmulation);
+    FEATURE_SUPPORT_GET(D3D12_RESOURCE_HEAP_TIER, m_dOptions, ResourceHeapTier);
+
+    // 1: Architecture (Deprecated)
```
The old Architecture feature seems deprecated. Its members are moved to Architecture1.

jenatali replied:
If running on a (very) old version of D3D that doesn't have Architecture1, we should populate the appropriate Architecture1 fields by using the query for Architecture instead.

shawnhar replied:
Yep.

Sanithovski replied:
I see the actual implementation of CheckFeatureSupport is basically redirecting Architecture inquiries to Architecture1, so they're accessing the same pieces of data. Is it OK to query these fields only from Architecture, not from Architecture1? Also will we one day drop support for Architecture entirely (CheckFeatureSupport will no longer take in that enum)?

jenatali replied:
The older version will never be deprecated. There exist apps out there today that are calling it, and if those apps run on a newer OS, they'll still call it and it still needs to work.

I think what we probably want to do here is: use Architecture1 if it's available, but if it's not, fall back to use Architecture, and fill out the fields that aren't there to some reasonable default value.

Sanithovski commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,277 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+        return m_hStatus;
+    }
+
+    static std::variant<CD3DX12FeatureSupport, HRESULT> Create(ID3D12Device* pDevice)
```
This is the factory method as discussed in the team chat. It returns either the initialized FeatureSupport intance on success or the HRESULT error code on failure.

Sanithovski commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,277 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+    }
+
+    // 3: Feature Format Support
+    HRESULT FormatSupport(DXGI_FORMAT Format, D3D12_FORMAT_SUPPORT1& Support1, D3D12_FORMAT_SUPPORT2& Support2)
```
I kept the same call routine for the more complicated feature checks except that I replaced the input struct with flat function parameters. I kept the return type HRESULT since the check may return non-trivial errors (not caused by invalid parameters but due to certain format not supported). Probably not the best way to do so?

Sanithovski commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,277 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+
+    // 6: GPU Virtual Address Support
+    // TODO: Check if it's exactly the same field as the one in the D3D12_Options feature
+    // FEATURE_SUPPORT_GET(UINT, m_dGPUVASupport, MaxGPUVirtualAddressBitsPerResource);
```
There is an entry with the same name in D3D12_Options feature. The two seem to be pointing to the same data field, but I need to doublecheck to make sure they're exactly the same thing.

Sanithovski commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,277 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+    return FEATURE.OPTION; \
+}
+
+#define FEATURE_SUPPORT_GET_NAME(RETTYPE,FEATURE,OPTION,NAME) \
```
Another macro in case where the getter function has to have a different name from the feature.

Sanithovski commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,277 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+//------------------------------------------------------------------------------------------------
+// Peter: Implementation for the new CheckFeatureSupport API
+
+#define FEATURE_SUPPORT_GET(RETTYPE,FEATURE,OPTION) \
```
Macros to set up an getter function for each entry in the feature data. Right now the getter function has the same name as the feature entry (if the entry is called "SomeFeatureSupported" then the getter function is also "SomeFeatureSupported()").

jenatali replied:
I'm personally not a huge fan of macros like this, since they only save 1 repetition and only save typing `() const { return` and `}`, but I don't feel super strongly about it. What I do feel strongly about here is that we shouldn't leak these macros into user code, so if we keep them, they need to be `#undef`'d at the end of the header.

vdwtanner replied:
Good use of the macros :) The GitHub comments were a great help understanding what each one did without having to see them in use. Thoughts on adding them as comments in the code too?

vdwtanner replied:
Oops, looks like the page hadn't loaded Jesse's comments before I replied. I'll defer to him here.

Sanithovski replied:
I'm using the macros so that if I need to change how similar functions are implemented, I can change it in one place instead of going over the code and changing all of them. This may save some time in the development phase, but indeed it's achieved at the cost of code readability. I'll probably keep the macros until the design patterns are finalized, and by that time I'll start replacing unnecessary macros with code. Does that sounds good?

Sanithovski replied:
I'll add the comments on what these macros do to the code as well. Thanks for the advice!

Sanithovski commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,277 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+    return FEATURE.OPTION; \
+}
+
+#define FEATURE_SUPPORT_GET_NODE_INDEXED(RETTYPE,FEATURE,OPTION) \
```
Macro to set up getter functions for features indexed by NodeIndex. Default to 0, or the first available graphics device node.

Sanithovski commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,277 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+        // Initialize static feature support data structures
+        INITIALIZE_MEMBER_DATA_CHECKED(D3D12_FEATURE_D3D12_OPTIONS, m_dOptions);
+        INITIALIZE_MEMBER_DATA_CHECKED(D3D12_FEATURE_GPU_VIRTUAL_ADDRESS_SUPPORT, m_dGPUVASupport);
+        m_dShaderModel.HighestShaderModel = D3D_HIGHEST_SHADER_MODEL; // TODO: What to do if highest model is inconsistent in the DLL and in the header?
```
I ran into a problem where the d3d12.h header in the repo has more entries than the D3D12 runtime on my local machine. In this case the D3D_HIGHEST_SHADER_MODEL is not recognized and the initialization would fail. I can think of several ways to deal with this issue:
1. Try out every enum entry in reverse order until one version is supported. Store that version entry.
2. Do not initialize the data field during construction. Let the user provide the version they want in the getter API.

jenatali replied:
Yeah, option 1 is the correct option here, and is one of the primary pieces of value that this helper can provide.

shawnhar replied:
Nice!  This is exactly the case that we will need to handle for when a game using this helper is run on an older version of Windows.  Trying them in reverse order seems like the right thing to me.

Sanithovski replied:
Thanks for the advice! I guess that's how users can feel comfortable using the new APIs. :)

Sanithovski commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,277 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+    FEATURE_SUPPORT_GET(UINT, m_dGPUVASupport, MaxGPUVirtualAddressBitsPerProcess);
+
+    // 7: Shader Model
+    D3D_SHADER_MODEL HighestShaderModel(D3D_SHADER_MODEL RequestedShaderModel = D3D_HIGHEST_SHADER_MODEL)
```
The API can also return a BOOL indicating whether the requested version is supported or not. Maybe it's more straightforward for the users? I think it depends on how the user will be using the API -- whether they more often want to know the highest supported version, or just whether a specific version is supported. Or maybe we can have both?

Sanithovski commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,277 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+    return FEATURE[NodeIndex].OPTION; \
+}
+
+#define INITIALIZE_MEMBER_DATA_CHECKED(FEATURE, MEMBER) \
```
Used when initializing a member in the Init() function. It will terminate the initialization process when there's an error.

Sanithovski commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,277 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+class CD3DX12FeatureSupport
+{
+public:
+    CD3DX12FeatureSupport(ID3D12Device* pDevice)
```
I'm keeping both the constructor and the factory pattern here to show what they look like. Both are working fine right now.

jenatali replied:
I think we should add a default constructor which sets `m_pDevice` to null and `m_hStatus` to a specific failure code which indicates that `Init` hasn't been called yet.

Sanithovski replied:
Will do! By the way, which error code would you suggest using for default initialization? I was thinking E_INVALIDARG but it's the same code with most other errors that can arise from CheckFeatureSupport.

jenatali replied:
`E_NOTIMPL` is a reasonable choice, or `E_UNEXPECTED`.

Sanithovski reviewed (commented):
I pointed out some code snippets for potential feedback, but you're welcome to leave feedback anywhere, anytime. It's easier to view them in the "Files Changed" tab so you can see the context.

jenatali commented on include/directx/d3dx12.h:
```diff
@@ -3293,6 +3293,7 @@ inline bool operator==( const D3D12_RENDER_PASS_DEPTH_STENCIL_DESC &a, const D3D
 #define D3DX12_COM_PTR_GET(x) x.p
 #define D3DX12_COM_PTR_ADDRESSOF(x) &x.p
 #endif
+#include <variant>
```
`<variant>` is C++17. We should only pull this in if `__cplusplus` is >= 201703L, and we should probably also add a preprocessor macro that users can set to opt out of including this, since `<variant>` is a pretty heavyweight header.

vdwtanner replied:
That would cause a different public API surface depending on whether or not variant is included. That inconsistency seems less than ideal, right? Would it be better to just use a different approach in this case, such as returning an HRESULT and returning the FeatureSupport class via out param?

jenatali replied:
That seems okay to me, since this is header-only. If you want to use a `std::variant`-based API, then do so. If you don't want the bloat required to do that, then opt out.

shawnhar replied:
Looks like this still needs a preprocessor method for users to opt-out of using <variant>?

Sanithovski replied:
Resolved using macros. I used `__cplusplus` and `_MSVC_LANG` for identifying the version of C++ standard, the latter of which may only work in Visual studio. The user also need to define a specific flag in order to use the <variant> option. If it's not turned on, the Create() factory method will return only the feature, not the HRESULT. Users can still know whether the initialization succeeds by using GetStatus() to access the error code.

jenatali commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,277 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+        m_hStatus = Init();
+    }
+
+    HRESULT Init()
```
1. I think `Init()` should take the `ID3D12Device*` instead of assuming it's already set in the class.
2. It seems odd that `Init()` returns the `HRESULT`, but doesn't set it on the class, only the constructor sets it on the class.

Sanithovski replied:
Thanks for pointing out! I thought I will only use m_hStatus in the constructor design, so I forgot to add it to the factory design pattern. I've added this to the next update.

jenatali commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,277 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+        m_dRootSignature.HighestVersion = D3D_HIGHEST_ROOT_SIGNATURE_VERSION;
+        INITIALIZE_MEMBER_DATA_CHECKED(D3D12_FEATURE_ROOT_SIGNATURE, m_dRootSignature);
+        INITIALIZE_MEMBER_DATA_CHECKED(D3D12_FEATURE_D3D12_OPTIONS2, m_dOptions2);
+        INITIALIZE_MEMBER_DATA_CHECKED(D3D12_FEATURE_SHADER_CACHE, m_dShaderCache);
```
Instead of using an early-out error approach for some of these, e.g. the shader cache flags, we should more cleanly handle the case where they're not supported by the D3D runtime.

Sanithovski replied:
I think CheckFeatureSupport won't return an error even if a certain feature exists but is not supported by the runtime. It will fail, however, if the runtime doesn't know about the feature at all, for example the runtime has no D3D12_FEATURE_SHADER_CACHE in its feature enum. I think this is the issue you're referring to? This is indeed one of the goals of this project (which I haven't achieved yet...). 

One question are the features ordered in incremental order, i.e., is it true that newer features will always have a higher feature enum assigned to them? And is it true that if a newer feature is supported, an older feature would always be supported? Here by supported I mean the D3D12 runtime can recognize the feature enum when checking it. This can impact how I'm going to resolve this issue.

shawnhar replied:
In practice it is /probably/ true that feature enum values always increase over time, but let's not rely on that.  It's a fragile assumption that might not always remain true.  I think we should design this to handle any combination of features not being available in any order.

Generally I think the principle should be, if a feature support check fails by returning that D3D didn't recognize that feature (as opposed to any other kind of failure) then don't bubble that failure code out to the caller of the API, but instead just fill in default "not supported" values for whatever feature was being queried at the time.

Sanithovski replied:
I modified the code such that unsupported feature structs will be initialized to default/invalid values instead of failing.

jenatali commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,277 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+    // 2: Feature Levels
+    // User inputs a list of feature levels. The function returns the highest level supported.
+    // Option 1: Straightforward implementation. Only perform forwarding
+    HRESULT HighestFeatureLevel(UINT NumFeatureLevels, const D3D_FEATURE_LEVEL* pFeatureLevelRequested, D3D_FEATURE_LEVEL& MaxSupportedFeatureLevel)
```
Personally I think this should be cached during creation, and just find the highest feature level supported from the full list available in the header.

shawnhar commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,277 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+
+    HRESULT Init()
+    {
+        // TODO: Think about hot to catch and report errors
```
One option would be to let the user decide this.  We could make it so that, for instance, if you include the header normally it will have a constructor that just zero-initializes all the data followed by separate Init method, but if the user does something like:

#define CD3DX12_FEATURE_SUPPORT_USE_EXCEPTIONS

before including d3dx12.h, then the constructor would actually do the initialization work and throw an exception if that failed.  This way only developers who like to use exceptions would have the exception code path included at all.

Sanithovski replied:
The good news is that if we're filling in default values for features that we fail to check, we probably don't need to throw errors. It is expected that some runtime doesn't even know that a D3D12_FEATURE entry exists, but we probably don't want to return immediately but should continue to check for the remaining features.

There are some checks that may cause errors. See the thread below on whether these features (Shader model, feature level and root signature) may throw errors.

Sanithovski replied:
Other features may introduce complexities in this process. I'll follow up this discussion in the PR for the second part of feature checks. There we can discuss how to handle the exceptions with the full Init() function visible to us.

Sanithovski replied:
Forwarded to next PR.

shawnhar commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,277 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+    // 2: Feature Levels
+    // User inputs a list of feature levels. The function returns the highest level supported.
+    // Option 1: Straightforward implementation. Only perform forwarding
+    HRESULT HighestFeatureLevel(UINT NumFeatureLevels, const D3D_FEATURE_LEVEL* pFeatureLevelRequested, D3D_FEATURE_LEVEL& MaxSupportedFeatureLevel)
```
I wonder if it would be useful to have a version of HighestFeatureLevel where the user doesn't have to pass in anything at all, and the helper just returns the highest one that is supported?

Sanithovski replied:
I think I've simplified this function in the latest update. The function now takes no parameters but simply returns the highest supported feature level.

Sanithovski replied:
Resolved. Only returns highest feature level.

shawnhar commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,277 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+    }
+
+    // 3: Feature Format Support
+    HRESULT FormatSupport(DXGI_FORMAT Format, D3D12_FORMAT_SUPPORT1& Support1, D3D12_FORMAT_SUPPORT2& Support2)
```
FormatSupport, MultisampleQualityLevels, and FormatInfo are similar in that they all take in a DXGI_FORMAT, and return different information about that format.  Could/should they be combined?

Sanithovski replied:
It seems MultisampleQualityLevels require user to provide SampleCount and Flags in addition to the format. I'm not sure it should be combined with the other two. FormatSupport and FormatInfo, on the other hand, seem to be easier to combine.

One thing I noticed is that FormatSupport has 2 output parameters, and I think it's making the API more complicated to use. If we combine these feature checks there will be more output parameters. Any advice for resolving this issue?

Sanithovski replied:
Forwarded to the next PR.

shawnhar commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,277 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+    }
+
+    // 3: Feature Format Support
+    HRESULT FormatSupport(DXGI_FORMAT Format, D3D12_FORMAT_SUPPORT1& Support1, D3D12_FORMAT_SUPPORT2& Support2)
```
Style question: should these complicated methods that have output parameters use &, or *?

The advantage of & is that it is not possible for someone to pass NULL.

But the advantage of * is that it makes it obvious at the call site that the function will be writing to the value.  References are invisible at the call site which can sometimes make code less readable.

Sanithovski replied:
Another option is to return them instead of using output parameters. In the case of FormatSupport we can return a std::pair, which I think shouldn't be more complicated for developers than using output parameters.

The question is does the HRESULT from these functions matter? I see that FormatSupport may return E_FAIL if the format is not supported, but there are also empty flags in D3D12_FORMAT_SUPPORT1 and D3D12_FORMAT_SUPPORT2 that could mean the same thing. Can we ignore the HRESULT and return these empty flags instead?

Sanithovski replied:
Forwarded to next PR.

shawnhar commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,277 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+    FEATURE_SUPPORT_GET(UINT, m_dGPUVASupport, MaxGPUVirtualAddressBitsPerProcess);
+
+    // 7: Shader Model
+    D3D_SHADER_MODEL HighestShaderModel(D3D_SHADER_MODEL RequestedShaderModel = D3D_HIGHEST_SHADER_MODEL)
```
Shader mode and feature level seem a bit similar (there's a set of them, more are added over time, and it's interesting to learn things like what is the highest one that's supported).  Consider making the available query methods follow the same pattern for both?

Sanithovski replied:
I think I've updated them in the previous commit. Now shader model and root signature will only return the highest shader model/root signature version to the user and don't require user inputs.

Sanithovski replied:
Resolved. Only returns the highest shader model. Same for root signature.
I forgot to put a "const" for this function definition. I've added it and it should come up in the next commit.

Sanithovski replied:
OK the const keyword is added.

shawnhar commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,277 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+
+    // Feature support data structs
+    D3D12_FEATURE_DATA_D3D12_OPTIONS m_dOptions;
+    std::vector<D3D12_FEATURE_DATA_ARCHITECTURE> m_dArchitecture; // Cat2 NodeIndex, should be deprecated
```
Note that std::vector is an allocating container, so it will allocate memory, throw exceptions if that allocation fails, and it's not just a simple copy to move structs that contain it.

This is probably ok for the purpose of this helper, but I wanted to point it out in case others have opinions.  If we really want to avoid internal allocations, std::array would avoid any dynamic allocation (at the cost of the size having to be fixed at compile time).

Sanithovski replied:
I would like to avoid any dynamic allocation as well, but unfortunately the size of these data structures depend on the number of computing nodes at runtime. If we use statically allocated arrays (std::array) we must predetermine a size that's large enough to accommodate any possible node count, which can take up too much space for most of the time since most computers only have one or two of these graphics nodes.

However, if there's an (not too high) upper limit on the number of graphics nodes, or if there's a way to determine that number at compile time, I will vote for a static array.

Sanithovski replied:
Forwarded to next PR.

Sanithovski commented:
It seems the additional comments in the commits were collapsed. Reposting here:

- Added comments to Macros
- Create a default constructor that sets device pointer to null and status to a predefined error code.
- Init function now takes in a pointer to the device
- Stores errors in m_hStatus during failed initialization
- Improved checking routines for feature checks requiring hints, including ShaderModel, RootSignature and FeatureLevels
- Add Undefs for macros at the end of class definition

shawnhar commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,370 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+    HRESULT Init(ID3D12Device* pDevice)
+    {
+        if (!pDevice) {
+            m_hStatus = E_NOTIMPL;
```
Should return E_INVALIDARG here, as the problem is an invalid parameter was passed to the function.

Sanithovski replied:
Resolved. Now returns E_INVALIDARG.

jenatali commented on include/directx/d3dx12.h:
```diff
@@ -4041,54 +4041,76 @@ class CD3DX12_NODE_MASK_SUBOBJECT
-    : m_pDevice(pDevice)
     {
-        m_hStatus = Init();
+        m_hStatus = Init(pDevice);
```
Now that you're assigning `m_hStatus` on every `CheckFeatureSupport` call, you don't need to re-assign it here.

jenatali commented on include/directx/d3dx12.h:
```diff
@@ -4217,13 +4239,7 @@ class CD3DX12FeatureSupport
-        D3D12_FEATURE_DATA_SHADER_MODEL dShaderModel;
-        m_pDevice->CheckFeatureSupport(D3D12_FEATURE_SHADER_MODEL, &dShaderModel, sizeof(D3D12_FEATURE_DATA_SHADER_MODEL));
-        return dShaderModel.HighestShaderModel;
+        return min(RequestedShaderModel, m_dShaderModel.HighestShaderModel);
```
The reason the D3D API for these uses an in-out pattern for this is because:
1. The app compiled with one set of enum values for feature level
2. D3D can be updated to include new enum values
3. If D3D returns new enum values to an app that doesn't understand those values, the app is seeing data that it doesn't necessarily know how to handle

For `d3dx12.h`, since this is compiled into the app, I think we can just use getters, instead of passing in a requested shader model. The value returned from the getter is *guaranteed* to be one of the values that the app knew about when it was compiled.

Sanithovski commented:
Made some revisions based on further reviews:
- Removed m_hStatus update in constructors.
- Simplified FEATURE_LEVEL checking. It will now return the highest feature level only.
- Removed input parameters from ROOT_SIGNATURE and SHADER_MODEL. They will now only return the highest version supported.
- Modified initialization procedure. Instead of quitting and fail with an error, the Init procedure will initialize data structs with default/invalid values when a feature is not supported.
- Updated Architecture feature checking to use old Architecture instead of Architecutre1 on older systems. (If a feature is shown as supported in Architecture but not Architecture1, the class will report the feature to be **supported**).

Sanithovski commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,432 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+            m_dOptions.OutputMergerLogicOp = false;
+            m_dOptions.MinPrecisionSupport = D3D12_SHADER_MIN_PRECISION_SUPPORT_NONE;
+            m_dOptions.TiledResourcesTier = D3D12_TILED_RESOURCES_TIER_NOT_SUPPORTED;
+            m_dOptions.ResourceBindingTier = (D3D12_RESOURCE_BINDING_TIER)0;
```
Some enums have an entry that stands for NOT_SUPPORTED and it's usually 0. Others don't, but I still set them to 0 if the entry is not occupied.

jenatali commented on include/directx/d3dx12.h:
```diff
@@ -4072,6 +4072,8 @@ if (FAILED(m_hStatus = m_pDevice->CheckFeatureSupport(FEATURE, &MEMBER, sizeof(M
     return m_hStatus; \
 }
 
+#define INITIALIZE_FAILED(FEATURE, MEMBER) FAILED(m_hStatus = m_pDevice->CheckFeatureSupport(FEATURE, &MEMBER, sizeof(MEMBER)))
```
I'm pretty sure you don't want to assign `m_hStatus` as part of this macro. Otherwise, the overall success/failure of the entire initialization process will just boil down to the last feature that was checked.

Sanithovski replied:
Oh! My bad. I didn't want to do this assignment! Thanks for pointing out.

Sanithovski commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,432 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+
+        
+        // Initialize static feature support data structures
+        if (INITIALIZE_FAILED(D3D12_FEATURE_D3D12_OPTIONS, m_dOptions)) {
```
Instead of failing, the struct corresponding to an unrecognized feature will be set to default/invalid values. 
Is there a faster way to set values for all the data fields? For example using `memset`?

jenatali replied:
`memset` is an option if the invalid values are all 0. You can also just use ` = {};` to default-initialize all values. I don't have a strong preference between these 3 options, as long as we first make sure that all of the defaults/invalid values are 0s and don't need to be -1s or other invalid values.

Sanithovski replied:
I'll see if I can find a good way to simplify it. For now I'll keep it this way for this PR.

Sanithovski commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,432 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+    }
+
+    BOOL CacheCoherentUMA(UINT NodeIndex = 0) {
+        return m_dArchitecture1[NodeIndex].CacheCoherentUMA || m_dArchitecture[NodeIndex].CacheCoherentUMA;
```
Returns true as long as one of the two Architecture struct returns true. It should work on older D3D12 without Architecture1.

jenatali commented on include/directx/d3dx12.h:
```diff
@@ -4114,18 +4152,35 @@ class CD3DX12FeatureSupport
+            }
+
+            m_dArchitecture1[i].NodeIndex = i;
+            if (INITIALIZE_FAILED(D3D12_FEATURE_ARCHITECTURE1, m_dArchitecture1[i])) {
```
I think the pattern I'd use here would be:

```c++
if (INITIALIZE_FAILED(D3D12_FEATURE_ARCHITECTURE1, m_dArchitecture1[i])) {
    D3D12_FEATURE_DATA_ARCHITECTURE localArch0 = {};
    localArch0.NodeIndex = i;
    if (INITIALIZE_FAILED(D3D12_FEATURE_ARCHITECTURE, localArch0)) {
        localArch0.TileBasedRenderer = false;
        localArch0.UMA = false;
        localArch0.CacheCoherentUMA = false;
    }

    m_dArchitecture1[i].TileBasedRenderer = localArch0.TileBasedRenderer;
    m_dArchitecture1[i].UMA = localArch0.UMA;
    m_dArchitecture1[i].CacheCoherentUMA = localArch0.CacheCoherentUMA;
    m_dArchitecture1[i].IsolatedMMU = false;
}
```

So that way, we only have one copy of the architecture data in the structure, and we don't need to query D3D for both architecture structs. If the 1 version is supported, it's a superset of the older version and that's all we need to do. If the 1 version's not supported, we can query the older version, and copy the relevant fields to the 1 version, and fill in defaults for the stuff that's only in the newer version.

Sanithovski replied:
Great idea! Updated.

amarpMSFT commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,432 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+    }
+
+    // 0: D3D12_OPTIONS
+    FEATURE_SUPPORT_GET(BOOL, m_dOptions, DoublePrecisionFloatShaderOps);
```
These macros are great for the implementation.  But they make it a bit harder to look at the class and see what methods can be called on it.  Perhaps all public members of the class could be predeclared at the top for the benefit of folks who don't wan't to bother sifting past implementation details?

Sanithovski replied:
For the Macros, I mainly used them to simplify implementation so if there's a change in the call pattern I don't need to go and change each line but instead I can just change the Macro. However, they do make the code hard to read. After the implementation is done, I will remove the macro and replace the call with actual code.

For the second part, do you mean I should put the public function declaration at the top and perform the actual implementation afterwards? Also does that mean the definition should take place outside the class declaration? I can certainly do that if it improves readability!

Sanithovski replied:
I'm going to reorganize the code when I finalize the second part of the feature and before merging the code into the main branch. I will forward this comment to the new PR. Thanks!

shawnhar commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,421 @@ class CD3DX12_NODE_MASK_SUBOBJECT
 };
 
+//------------------------------------------------------------------------------------------------
+// Peter: Implementation for the new CheckFeatureSupport API
```
Best practice is not to include author names in comments, as this won't mean anything to people not on the team and soon becomes obsolete as people move around and teams change.  The author can always be found in git history if needed.

Sanithovski replied:
Resolved. Removed my name.

shawnhar commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,421 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+        HRESULT result;
+
+        // SHADER_MODEL_6_X
+        for (int i = D3D_SHADER_MODEL_6_7; i >= D3D_SHADER_MODEL_6_0; i--) {
```
I wonder if using looping over a table of known shader models here (like the pattern you used in QueryHighestFeatureLevel) could simplify this code by avoiding the repetition for SHADER_MODEL_6_X vs. SHADER_MODEL_5_1 handling?

Sanithovski replied:
The two routines are slightly different. The feature level data requires a list of feature levels to be checked in on CheckFeatureSupport call, so I have to create a list and send all possible levels in at a single time. The shader model data only takes in one shader model at a time, which requires a loop. I think creating a list can't save many lines of code, but I believe it will be more readable for the developers (and more maintainable as well). I'll make the update.

Sanithovski replied:
Resolved. SHADER_MODEL now uses the same check routine as FEATURE_LEVEL. I also updated the routine for ROOT_SIGNATURE so the three are consistent.

shawnhar commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,421 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+
+    // Feature support data structs
+    D3D12_FEATURE_DATA_D3D12_OPTIONS m_dOptions;
+    // std::vector<D3D12_FEATURE_DATA_ARCHITECTURE> m_dArchitecture; // Cat2 NodeIndex, should be deprecated
```
I'd suggest removing these commented out entries.

Sanithovski replied:
I left them in there to remind me the entries in the feature enum. I can remove them now that the implementation is done.

Sanithovski replied:
Actually there are several places where I created a feature data struct in the function instead of using these commented out member struct. I guess it's better to use these member structs to avoid repeated allocation? See FormatSupport for an example.

shawnhar replied:
I actually think it's better to create these just where you need them.  That will be very cheap / free stack allocation, only lasting as long as the function that uses it, and makes clear that the usage of the object is only temporary.

Sanithovski replied:
Resolved. Used temporary data structs and removed unused member structs.

shawnhar commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,421 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+
+    CD3DX12FeatureSupport(ID3D12Device* pDevice)
+    {
+        Init(pDevice);
```
If Init() fails, this could leave some of the data fields uninitialized.  Depending on what the desired design is, either throw an exception on error, or make sure all the member data is set to well-defined default values in case of failure.  Or we could just remove this constructor?

Sanithovski replied:
At this time, only three feature checks -- shader model, root signature, and feature level -- may cause an error that makes Init() fail. All other initialization routines will not cause failure. 

I checked the current implementation. It seems shader model and root signature will only return E_INVALIDARG when it doesn't recognize the feature, which is handled in Init(). Feature level may return E_DXGI_NOT_SUPPORTED only when non of the requested feature level is supported, while Init() will pass in a list of all feature levels available in d3d12.h. 

So looks like the current init() will not fail unless 1) There is a D3D12 that supports none of the feature levels, or 2) An error other than the above can rise even if it's not written in CheckFeatureSupport or relative functions. If these cases won't happen, we probably don't need to worry about failures.

Sanithovski replied:
Updating from our Teams discussion: Some of the features may still return a specific error code not handled by Init(). New features and implementation may also introduce unexpected errors. So it's better to keep the error reporting mechanism in the class as well as in Init().

I'll follow up this discussion in the PR for the second part of feature checks. There we can discuss how to handle the exceptions with the full Init() function visible to us.

shawnhar reviewed (changes requested):
This is looking great!  I added a few minor comments to this version.   Also I notice there are still a few Pending comments left from code review on earlier versions.  Let's get these all resolved before completing the PR.  If you disagree with any of the feedback, feel free to say so as a reply to it and then resolve the items!  Or if you aren't sure how to proceed, let's use the Teams chat channel to discuss any items that aren't yet settled.

One earlier feedback that I think got missed was about <variant> being heavyweight and not something every customer is going to want.  I'd suggest creating a preprocessor symbol that lets customer enable or disable that part of the API as they see fit.  Perhaps wrap the include of <variant> and the Create method with something like #ifdef D3DX12_USE_VARIANT?  This way that heavyweight functionality is not included by default, yet anyone who wants it can just #define D3DX12_USE_VARIANT before they #include "d3dx12.h".

alexgong14 commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,421 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+        for (int i = D3D_SHADER_MODEL_6_7; i >= D3D_SHADER_MODEL_6_0; i--) {
+            m_dShaderModel.HighestShaderModel = static_cast<D3D_SHADER_MODEL>(i);
+            result = m_pDevice->CheckFeatureSupport(D3D12_FEATURE_SHADER_MODEL, &m_dShaderModel, sizeof(D3D12_FEATURE_DATA_SHADER_MODEL));
+            if (SUCCEEDED(result) || result != E_INVALIDARG) {
```
maybe simplify this to if (result != E_INVALIDARG)?

Sanithovski replied:
Resolved

alexgong14 commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,421 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+        for (int i = D3D_ROOT_SIGNATURE_VERSION_1_1; i >= D3D_ROOT_SIGNATURE_VERSION_1; i--) {
+            m_dRootSignature.HighestVersion = static_cast<D3D_ROOT_SIGNATURE_VERSION>(i);
+            result = m_pDevice->CheckFeatureSupport(D3D12_FEATURE_ROOT_SIGNATURE, &m_dRootSignature, sizeof(D3D12_FEATURE_DATA_ROOT_SIGNATURE));
+            if (SUCCEEDED(result) || result != E_INVALIDARG) {
```
here too, can simplify to result != E_INVALIDARG

Sanithovski replied:
Resolved

Sanithovski commented on include/directx/d3dx12.h:
```diff
@@ -4037,6 +4038,421 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+
+        HRESULT result = m_pDevice->CheckFeatureSupport(D3D12_FEATURE_MULTISAMPLE_QUALITY_LEVELS, &dMultisampleQualityLevels, sizeof(D3D12_FEATURE_DATA_MULTISAMPLE_QUALITY_LEVELS));
+        
+        NumQualityLevels = dMultisampleQualityLevels.NumQualityLevels;
```
Zichuan pointed out that the out parameter, NumQualityLevels, should be set to 0 upon failure. I checked the source code for this case. It seems that CheckFeatureSupport will set NumQualityLevels to 0 if there's an unexpected failure (and return E_FAIL), so this should be the correct behavior. Thanks, Zichuan!

Sanithovski replied:
Though 0 is used as a default fallback value, it can also be produced by a successful CheckFeatureSupport call that hasn't run into issues. This suggests users must check the HRESULT returned to determine whether the check succeeds.

Sanithovski commented:
Updated code based on the third round of code reviews:
- Removed my name from the comments
- Use E_INVALIDARG instead of E_NOTIMPL for default error message
- Initialize NumQualityLevels to 0 upon failure
- Changed design pattern for SHADER_MODEL and ROOT_SIGNATURE_VERSION checks
- Removed unnecessary member structs

Sanithovski commented:
Added preprocessor statements to allow conditional use of `<variant>`

Sanithovski commented:
I'm keeping this PR open for review while posting my PR for the second part of the API separately. I've been constantly syncing the two branches so the other branch also receives the updates from this PR. I will close and merge this PR once everyone approves the current changes.
